packages/@expo/cli/src/localModules/generation.ts

```diff
@@ -0,0 +1,362 @@
+    for (const glob of excludePathsGlobs(projectRoot)) {
+      if (path.matchesGlob(absoluteDirPath, glob)) {
+        return;
+      }
+    }
+
+    const dir = fs.opendirSync(absoluteDirPath);
+    for await (const dirent of dir) {
+      const absoluteDirentPath = path.resolve(absoluteDirPath, dirent.name);
+      if (
+        dirent.isFile() &&
+        /\.(kt|swift)$/.test(dirent.name) &&
+        fileSupposedToBeWatched(projectRoot, absoluteDirentPath)
+      ) {
+        addNewFile(projectRoot, absoluteDirentPath, filesWatched);
+      } else if (dirent.isDirectory()) {
+        await generateExportsAndTypesForDirectory(absoluteDirentPath);
+      }
+    }
+  };
+  await generateExportsAndTypesForDirectory(projectRoot);
+}
+
+function excludePathsGlobs(projectRoot: string): string[] {
+  return [
+    path.resolve(projectRoot, '.expo'),
+    path.resolve(projectRoot, '.expo', './**'),
+    path.resolve(projectRoot, '.expo', './**/*'),
+    path.resolve(projectRoot, 'node_modules'),
+    path.resolve(projectRoot, 'node_modules', './**'),
+    path.resolve(projectRoot, 'node_modules', './**/*'),
+    path.resolve(projectRoot, 'android'),
+    path.resolve(projectRoot, 'localModules'),
+    path.resolve(projectRoot, 'localModules', './**'),
+    path.resolve(projectRoot, 'localModules', './**/*'),
+    path.resolve(projectRoot, 'android'),
+    path.resolve(projectRoot, 'android', './**'),
+    path.resolve(projectRoot, 'android', './**/*'),
+    path.resolve(projectRoot, 'ios'),
+    path.resolve(projectRoot, 'ios', './**'),
+    path.resolve(projectRoot, 'ios', './**/*'),
+    path.resolve(projectRoot, 'modules'),
+    path.resolve(projectRoot, 'modules', './**'),
+    path.resolve(projectRoot, 'modules', './**/*'),
+  ];
+}
+
+export async function startModuleGenerationAsync({
+  projectRoot,
+  metro,
+}: ModuleGenerationArguments) {
+  const dotExpoDir = ensureDotExpoProjectDirectoryInitialized(projectRoot);
+  const { exp } = getConfig(projectRoot);
+  const filesWatched = new Set<string>();
+
+  const fileExcluded = (absolutePath: string) => {
+    for (const glob of excludePathsGlobs(projectRoot)) {
+      if (path.matchesGlob(absolutePath, glob)) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  createFreshMirrorDirectories(projectRoot);
+
+  process.env.EXPO_ROUTER_APP_ROOT = path.join(
+    projectRoot,
+    getRouterDirectoryModuleIdWithManifest(projectRoot, exp)
+  );
+
+  const removeFileAndEmptyDirectories = (absoluteFilePath: string) => {
+    if (fs.lstatSync(absoluteFilePath).isSymbolicLink()) {
+      fs.unlinkSync(absoluteFilePath);
+    } else {
+      fs.rmSync(absoluteFilePath);
+    }
+    let dirNow: string = path.dirname(absoluteFilePath);
+    while (fs.readdirSync(dirNow).length === 0 && dirNow !== dotExpoDir) {
+      fs.rmdirSync(dirNow);
+      dirNow = path.dirname(dirNow);
+    }
+  };
+
+  const onRemoveAppFile = (absoluteFilePath: string) => {
```

Comment:
this name indicates you dropped the `app` directory for me
```suggestion
  const onSourceFileRemoved = (absoluteFilePath: string) => {
```